### PR TITLE
GH-40602: [C++] Move mold linker flags to variables

### PR DESCRIPTION
**Pull request by cryos**
### Rationale for this change

Move the mold linker flags to the CMake `CMAKE_*_LINKER_FLAGS` variables. This is the more correct place to add them as they will only be used during linking, it would invalidate any [s]ccache less frequently if changing linkers and I think it is more clear/correct.

### What changes are included in this PR?

Moving the mold linker flags from the `CMAKE_*_FLAGS` to `CMAKE_*_LINKER_FLAGS`.

### Are these changes tested?

Existing tests and validation would cover these changes and the result is essentially the same as before.

### Are there any user-facing changes?

No
* GitHub Issue: #40602

### CODE REVIEW

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 1]
```diff
@@ -629,23 +629,27 @@ if(NOT WIN32 AND NOT APPLE)
...
+        set(MOLD_LINKER_FLAGS "--ld-path=${LD_MOLD}")
         message(STATUS "Using optional mold linker")
       else()
         message(STATUS "Using the default linker because compiler doesn't support mold: ${CMAKE_CXX_COMPILER_ID}"
         )
       endif()
+      if(MOLD_LINKER_FLAGS)
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${MOLD_LINKER_FLAGS}")
+        set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} ${MOLD_LINKER_FLAGS}")
+        set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${MOLD_LINKER_FLAGS}")
```
Comment (kou) [inline cpp/cmake_modules/SetupCxxFlags.cmake:651]:
```suggestion
        string(APPEND CMAKE_EXE_LINKER_FLAGS " ${MOLD_LINKER_FLAGS}")
        string(APPEND CMAKE_MODULE_LINKER_FLAGS " ${MOLD_LINKER_FLAGS}")
        string(APPEND CMAKE_SHARED_LINKER_FLAGS " ${MOLD_LINKER_FLAGS}")
```

**Review by kou — APPROVED** [review]
+1

Thanks. I couldn't remember these flags when I wrote this feature.

**Comment by cryos** [comment]
Thanks, I need to use `string(APPEND ...)` more, I had forgotten that had been added and love removing repetition!

**Comment by kou** [comment]
Merged.

BTW, do you want to work on related issue https://github.com/apache/arrow/issues/40400 ?

**Comment by cryos** [comment]
> Merged.
> 
> BTW, do you want to work on related issue https://github.com/apache/arrow/issues/40400 ?

I could help with that, thanks @kou

**Pull request merged**